src/libtls/tls_server.c

```diff
@@ -282,18 +385,54 @@ static status_t process_client_hello(private_tls_server_t *this,
 	}
 	rng->destroy(rng);
 
-	if (!this->tls->set_version(this->tls, version, version))
+	/* Parse supported TLS versions by peer */
+	if (versions.len != 0)
 	{
-		DBG1(DBG_TLS, "negotiated version %N not supported",
-			 tls_version_names, version);
-		this->alert->add(this->alert, TLS_FATAL, TLS_PROTOCOL_VERSION);
-		return NEED_MORE;
+		bio_reader_t *client_versions = bio_reader_create(versions);
+		while (client_versions->remaining(client_versions))
+		{
+			if (client_versions->read_uint16(client_versions, &version))
+			{
+				if (this->tls->set_version(this->tls, version, version))
+				{
+					DBG1(DBG_TLS, "negotiate version %N ",
+						 tls_version_names, version);
+					this->client_version = version;
+					break;
+				}
+			}
+		}
+		client_versions->destroy(client_versions);
+		if (this->client_version == 0)
```

Comment:
```suggestion
		if (!this->client_version)
```

I think you could move this block after the `if (versions.len != 0)` block (see below for details).